### ISSUE TITLE
[minillm] is it possible to split a huge model into 3 or 5 parts instead of 4?

### ISSUE THREAD

**Issue by chentianyangWHU** [microsoft/LMOps#65]
I noticed that the minillm algorithm usually splits a whole checkpoint into 4 equal parts using tools/convert_mp.py when dealing with larger LLM models like llama13B. During loading, each of the 4 A100 GPUs is assigned to load one of these parts using DeepSpeed. However, I currently only have 3 A100 GPUs, and it's clear that llama's vocab_size=32k cannot be evenly divided by 3, and some intermediate variables like hidden_size may also not be divisible by 3. My question is, can I split the whole checkpoint into uneven parts for the 3 GPUs (e.g., making 2 parts equal and the third part smaller), so that training can still be accomplished on the 3 GPUs (assuming sufficient memory)? If so, would the training code need to be modified in certain places? Thank you!

**Comment by chentianyangWHU** [comment]
Another question that arises from this is, if I have 15 A10 GPUs (24G), can I use the same method mentioned above to split llama13B and llama7B into 15 parts each (since the lengths of certain layers in the models may not be divisible by 15, it is likely that 14 of the parts will be of equal size and the remaining 1 part will be slightly larger or smaller), in order to train them on these 15 GPUs?

**Comment by t1101675** [comment]
It is hard to split the model using TensorParalell when the hidden state size is not divisible by the number of splits. The reason is that PyTorch assumes the tensor sizes on each model parallel process are the same when doing `all_gather` or `all_reduce`. 
I guess a better way to implement this may be **zero-padding** the hidden state size to a larger number divisible by the number of splits because adding several zeros after the hidden state does not affect the model outputs. However, this needs careful implementation to adjust the sizes of the model's weights by zero-padding.

**Comment by chentianyangWHU** [comment]
> It is hard to split the model using TensorParalell when the hidden state size is not divisible by the number of splits. The reason is that PyTorch assumes the tensor sizes on each model parallel process are the same when doing `all_gather` or `all_reduce`. I guess a better way to implement this may be **zero-padding** the hidden state size to a larger number divisible by the number of splits because adding several zeros after the hidden state does not affect the model outputs. However, this needs careful implementation to adjust the sizes of the model's weights by zero-padding.

thank you！
